llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp

```diff
@@ -2237,6 +2237,11 @@ InstructionCost VPWidenMemoryRecipe::computeCost(ElementCount VF,
     const Value *Ptr = getLoadStorePointerOperand(&Ingredient);
     assert(!Reverse &&
            "Inconsecutive memory access should not have the order.");
+    if ((isa<LoadInst>(Ingredient) &&
+         !Ctx.TTI.isLegalMaskedGather(Ty, Alignment)) ||
+        (isa<StoreInst>(Ingredient) &&
+         !Ctx.TTI.isLegalMaskedScatter(Ty, Alignment)))
+      return InstructionCost::getInvalid();
```

Comment:
shouldn't `getGatherScatterOpCost` return `Invalid` in those cases as we pass `IsMasked` ?

Reply:
I am not sure who has the responsibility to do the legal check transformations or TTI.
In RISCV, `getGatherScatterOpCost` will do the legal check and return invalid cost when it fail.
In the default target, `getGatherScatterOpCost` just return the constant cost no matter it is legal or not.

I found the legacy cost model explicit check if the `gather/scatter` is legal in `LoopVectorizationCostModel::setCostBasedWideningDecision()`.

So adding this check can prevent misaligned between legacy cost model and the vplan-based cost model when a target `isLegalMaskedGather() = false && getGatherScatterOpCost() is valid` .

Reply:
I see. I think the difference is that while the legacy cost model needs to consider all options for the input, no masked VPWidenMemoryRecipe should be created , if masked mem-ops aren't legal I think